pkg/groups/service.go

```diff
@@ -561,3 +565,335 @@ func NewService(ofga OpenFGAClientInterface, wpool pool.WorkerPoolInterface, tra
+	ctx, span := s.tracer.Start(ctx, "groups.V1Service.GetGroup")
+	defer span.End()
+
+	principal := authentication.PrincipalFromContext(ctx)
+	if principal == nil {
+		return nil, v1.NewAuthorizationError("unauthorized")
+	}
+
+	group, err := s.core.GetGroup(ctx, principal.Identifier(), groupId)
+	if err != nil {
+		return nil, v1.NewUnknownError(fmt.Sprintf("failed to get group %s for user %s: %v", groupId, principal.Identifier(), err))
+	}
+
+	if group == nil {
+		return nil, v1.NewNotFoundError(fmt.Sprintf("group %s not found", groupId))
+	}
+
+	return &resources.Group{
+		Id:   &group.ID,
+		Name: group.Name,
+	}, nil
+}
+
+// UpdateGroup updates the given resources.Group.
+//
+// Note: this is not implemented yet.
+func (s *V1Service) UpdateGroup(ctx context.Context, group *resources.Group) (*resources.Group, error) {
+	_, span := s.tracer.Start(ctx, "groups.V1Service.UpdateGroup")
+	defer span.End()
+
+	return nil, v1.NewNotImplementedError("service not implemented")
+}
+
+// DeleteGroup deletes a single group by its ID.
+func (s *V1Service) DeleteGroup(ctx context.Context, groupId string) (bool, error) {
+	ctx, span := s.tracer.Start(ctx, "groups.V1Service.DeleteGroup")
+	defer span.End()
+
+	principal := authentication.PrincipalFromContext(ctx)
+	if principal == nil {
+		return false, v1.NewAuthorizationError("unauthorized")
+	}
+
+	if err := s.core.DeleteGroup(ctx, groupId); err != nil {
+		return false, v1.NewUnknownError(fmt.Sprintf("failed to delete group %s for principal %s: %v", groupId, principal.Identifier(), err))
+	}
+
+	return true, nil
+}
+
+// GetGroupIdentities returns a page of resources.Identity associated with the given group.
+func (s *V1Service) GetGroupIdentities(ctx context.Context, groupId string, params *resources.GetGroupsItemIdentitiesParams) (*resources.PaginatedResponse[resources.Identity], error) {
+	ctx, span := s.tracer.Start(ctx, "groups.V1Service.GetGroupIdentities")
+	defer span.End()
+
+	paginator := types.NewTokenPaginator(s.tracer, s.logger)
+	if err := paginator.LoadFromString(ctx, *params.NextToken); err != nil {
+		s.logger.Error(fmt.Sprintf("failed to parse the page token: %v", err))
+	}
+
+	identities, pageToken, err := s.core.ListIdentities(ctx, groupId, *params.NextToken)
+	if err != nil {
+		return nil, v1.NewUnknownError(fmt.Sprintf("failed to list identities for group %s: %v", groupId, err))
+	}
+
+	paginator.SetToken(ctx, GROUP_TOKEN_KEY, pageToken)
+	metaParam, err := paginator.PaginationHeader(ctx)
+	if err != nil {
+		s.logger.Errorf("failed to create the pagination meta param: %v", err)
+		metaParam = ""
+	}
+
+	r := &resources.PaginatedResponse[resources.Identity]{
+		Meta: resources.ResponseMeta{Size: len(identities)},
+		Data: make([]resources.Identity, 0, len(identities)),
+		Next: resources.Next{PageToken: &metaParam},
+	}
+
+	for _, identity := range identities {
+		r.Data = append(r.Data, resources.Identity{Id: &identity})
```

Comment:
this might sound weird, but we might need to find a way to enhance these results with what we get from kratos, or at least find a way to add a link to the endpoint `getIdentity` that shows the right amount of data

i would argue that this for us is not a great endpoint due to the complexities around satisfying the aggregation of data (this would involve a call per identity object as i dont think kratos supports a "gimme all the identities in this filter")

Reply:
@shipperizer Yeah, I remember we talked about this. I tried something similar to `GetIdentity` in identity services. Let me know if this is OK at this moment so I can go ahead and hit the unit test cases.

For the identity details handling, I guess we will come to a consistent way when we know what fields should be filled in the response. (although we never know the identity schema that the user might use.)

```golang
func (s *V1Service) GetGroupIdentities(ctx context.Context, groupId string, params *resources.GetGroupsItemIdentitiesParams) (*resources.PaginatedResponse[resources.Identity], error) {
	ctx, span := s.tracer.Start(ctx, "groups.V1Service.GetGroupIdentities")
	defer span.End()

	paginator := types.NewTokenPaginator(s.tracer, s.logger)
	if err := paginator.LoadFromString(ctx, *params.NextToken); err != nil {
		s.logger.Error(fmt.Sprintf("failed to parse the page token: %v", err))
	}

	identities, pageToken, err := s.core.ListIdentities(ctx, groupId, *params.NextToken)
	if err != nil {
		return nil, v1.NewUnknownError(fmt.Sprintf("failed to list identities for group %s: %v", groupId, err))
	}

	paginator.SetToken(ctx, GROUP_TOKEN_KEY, pageToken)
	metaParam, err := paginator.PaginationHeader(ctx)
	if err != nil {
		s.logger.Errorf("failed to create the pagination meta param: %v", err)
		metaParam = ""
	}

	r := &resources.PaginatedResponse[resources.Identity]{
		Meta: resources.ResponseMeta{Size: len(identities)},
		Data: make([]resources.Identity, 0, len(identities)),
		Next: resources.Next{PageToken: &metaParam},
	}

	for _, user := range identities {
		userParts := strings.SplitN(user, ":", 2)
		identityID := userParts[1]

		kratosIdentity, _, err := s.kratos.GetIdentityExecute(
			s.kratos.GetIdentity(ctx, identityID),
		)
		if err != nil {
			s.logger.Warnf("skipping identity %s due to error: %v", user, err)
			continue
		}

		traits, ok := kratosIdentity.Traits.(map[string]string)
		if !ok {
			traits = make(map[string]string)
		}

		identity := &resources.Identity{
			Id:    &kratosIdentity.Id,
			Email: traits["email"],
		}

		fullName := traits["name"]
		surnameIndex := strings.LastIndex(fullName, " ")

		if surnameIndex > 0 {
			firstName := strings.TrimSpace(fullName[:surnameIndex])
			lastName := strings.TrimSpace(fullName[surnameIndex:])
			identity.FirstName = &firstName
			identity.LastName = &lastName
		}

		r.Data = append(r.Data, *identity)
	}

	r.Meta.Size = len(r.Data)
	return r, nil
}
```